### ISSUE TITLE
Bookmarks module

### ISSUE THREAD

**Issue by ColemanGariety** [breach/mod_strip#15]
I'd like to be able to store bookmarks in breach.

Does this module exist? If not, should it be a new module or integrated with mod_strip?

**Comment by spolu** [comment]
This should probably go in a separate module yes!

**Comment by jedahan** [comment]
I added a wiki page to keep track of modules / module idease, not sure if thats the best place, or just adding a tag for issues

**Comment by jonaldomo** [comment]
I would like to work on a bookmarks module over the weekend.  Has this been started?  Or can I start it?

**Comment by ColemanGariety** [comment]
@jonaldomo I was planning on starting it today, lemme see what I can get done on it and I'll post it here when I've started.

**Comment by jonaldomo** [comment]
@JacksonGariety I started posting my ideas at https://github.com/jonaldomo/mod_bookmark/wiki/Architecture.  I will start on it later tonight, let me know if you get anywhere.

**Comment by jedahan** [comment]
it might be useful to look at pinboard api https://pinboard.in/api/ and maybe be compat with that...

**Comment by spolu** [comment]
@jedahan that's a great idea to create a wiki page. Only it should probably be on github.com/breach/breach_core as `mod_strip` is one of the eventual modules on that wiki, no?

**Comment by spolu** [comment]
@jonaldomo Breach comes with a storage system (used for cookies and tabs) that could later on be turned distributed. 

It's called gig.fs, and available here: https://github.com/breach/gig.fs and exposed through the `breach_core` API here:
https://github.com/breach/breach_core/blob/master/lib/core_store.js

You can look at mod_strip's https://github.com/breach/mod_strip/blob/master/lib/tabs.js for an example

**Comment by julien-c** [comment]
It would be cool if you'd choose to implement classification of bookmarks via tags/labels, _NOT folders_. That would be an interesting different stance from Chrome.

**Comment by jedahan** [comment]
@spolu for the wiki page yes, now that I think it more carefully I was convoluting wiki pages with github issues. Moving now.

**Comment by spolu** [comment]
@jedahan thanks!

**Comment by joahg** [comment]
Just started looking into breach myself and started looking myself into creating a bookmarks mod using the [bit.ly](http://bit.ly) api. The upside of using the bit.ly api is that we would be able to create all the bookmarks by default as private, but make it easy for the user to share it, and take advantage of bitly's analytics platform for social bookmarking.

Also, bitly allows for "bundles", which should solve the problem of folders/tagging ;P

**Comment by spolu** [comment]
@JoahG sounds great, have you started a repo? have you update the module list page in the wiki?

**Comment by joahg** [comment]
@spolu well, yeah, I poked around with things (you can see my progress at [JoahG/mod_bookmarks](https://github.com/JoahG/mod_bookmarks), but I am unsure where to start with making a module for breach... What I have so far is what I was able to get from the [Creating a new module](https://github.com/breach/breach_core/wiki/Creating-a-new-module) wiki article. 

Once I grasp how the architecture of Breach modules is supposed to work, it shouldn't be hard to build it... Bitly provides comprehensible APIs that should make that connection fairly simple, and HTML/CSS styling for the bookmarks themselves shouldn't be too terribly difficult, my current issues are:
1. How would I go about adding a bar to the Breach module? (I have taken a look at the [mod_strip](https://github.com/breach/mod_strip) module, but haven't been able to figure that out yet)
2. How would I go about authenticating users through Bitly inside of the Breach environment, and once I do that, how would I store the returned API token?
3. How should the flow work through the module itself? As in, I get that I need to export an `init()` and a `kill()` method, but what else do I need to do?

Thanks

**Comment by spolu** [comment]
So you could start by adding context menu entries for bookmarking like "bookmark page" and "show bookmark". You can see how to add context menu in mod_strip's `lib/tabs.js`. This could be a good starting point.

As far as the UI to show the bookmarks, what do you have in mind? Do you want to display a full page? do you want it to be a drawer control on the left/right/bottom... The rest depends a lot on that.

As far as authenticating with bitly this is really up to you... you'd have to ask the users to signin first time they use the module I guess?

**Comment by jacksenechal** [comment]
It would be nice to be able to access bookmarks purely by the keyboard, filtering the list by what you type command-T style. It could be integrated into the URL bar in mod_strip, similar to Chrome or FF behavior, or it could have its own UI.

It would also be nice if the architecture had API connectors for various services, and those were all accessible through the same standard bookmarking UI. The user could choose to use bitly, pinboard, whatever, but the UX would stay basically the same.

**Comment by spolu** [comment]
Yep that'd be pretty cool indeed! Have a uniform UX search based on top of existing services would be a great way to bootstrap usage.
